### PR TITLE
[UDAF] Vectorization support remote udaf #10683

### LINKED ISSUE #10683: [UDAF] Vectorization support  remote  udaf function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

1.vectorization  support remote udaf  function

#### 1.create a remote udaf  for example to impl sum function 
```
CREATE AGGREGATE FUNCTION  rpc_add(int) RETURNS int PROPERTIES (
    "TYPE"="RPC",
    "OBJECT_FILE"="127.0.0.1:9000",
    "update_fn"="rpc_update",
    "merge_fn"="rpc_merge",
    "finalize_fn"="rpc_finalize"
);
```

#### 2. impl the fn interface and start a rpc server  
image

#### 3.use remote udaf function in  vectorization mode
image

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by chenlinzhong**
# Proposed changes

Issue Number: close #10683

## Problem Summary:
Currently, vectorized udaf can only be implemented locally based on c++ or Java. There are some problems need to be considered:

- The cost for getting started is high
- The requirements for the implementation quality of functions are relatively high, otherwise it will affect the stability of be, such as leading to core

To solove this problem , I think  the remote udaf function  is a good way  it can benifits:

- the cost for getting started is low, for it can use  any language to impl  udaf interface
- decouple with doris kernel， it good for be stability

We need to consider the performance, and whether the remote call will have a lot of overhead， Here I did a test
#### Step1. deploy with 2 BE node,  Table scheam:
```
CREATE TABLE `table1` (
  `event_day` date NULL,
  `siteid` int(11) NULL DEFAULT "10",
  `citycode` smallint(6) NULL,
  `username` varchar(32) NULL DEFAULT "",
  `pv` bigint(20) SUM NULL DEFAULT "0"
) ENGINE=OLAP
AGGREGATE KEY(`event_day`, `siteid`, `citycode`, `username`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`siteid`) BUCKETS 10
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
)
```

#### Step2 create remote udaf 
```
CREATE AGGREGATE FUNCTION  rpc_myadd(int) RETURNS int PROPERTIES (
    "TYPE"="RPC",
    "OBJECT_FILE"="127.0.0.1:9000",
    "update_fn"="rpc_sum_update",
    "merge_fn"="rpc_sum_merge",
    "finalize_fn"="rpc_sum_finalize"
);
```
#### Result
image

**we can concluded:**
- RPC udaf performance is better than that of non vectorized aggregate function by 6% - 29%
- The performance of RPC udaf is significantly lower than that of native vectorized aggregate functions, but when the number of data rows is 1000W, it can still return in seconds, and theoretically it can support most business scenarios

image
image
image
image

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
5. Has document been added or modified: (Yes/No/No Need)
6. Does it need to update dependencies: (Yes/No)
7. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yangzhg** [comment]
please add example to docs

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 1]
```diff
@@ -0,0 +1,762 @@
...
+#include "vec/core/block.h"
+#include "vec/core/column_numbers.h"
+#include "vec/core/field.h"
+#include "vec/core/types.h"
+#include "vec/data_types/data_type_string.h"
+#include "vec/io/io_helper.h"
+namespace doris::vectorized {
+
+template <bool nullable>
+void convert_col_to_pvalue(const vectorized::ColumnPtr& column,
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:46]: move those function to be/src/exprs/rpc_fn.cpp

File: samples/doris-demo/remote-udf-cpp-demo/cpp_function_service_demo.cpp [hunk 2]
```diff
@@ -128,4 +128,4 @@ int main(int argc, char** argv) {
     // Wait until Ctrl-C is pressed, then Stop() and Join() the server.
     server.RunUntilAskedToQuit();
     return 0;
-}
+}
```
Comment (yangzhg) [inline samples/doris-demo/remote-udf-cpp-demo/cpp_function_service_demo.cpp:131]: reset to  origin

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 3]
```diff
@@ -0,0 +1,762 @@
...
+        const auto* col = vectorized::check_and_get_column<vectorized::ColumnUInt8>(null_col);
+        auto& data = col->get_data();
+        null_map->Add(data.begin() + start, data.begin() + end);
+        convert_col_to_pvalue<true>(column, data_type, arg, start, end);
+    } else {
+        convert_col_to_pvalue<false>(column, data_type, arg, start, end);
+    }
+}
+
+#define MAX_BUFFERED_ROWS 4096
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:465]:
```suggestion
constexpr int64_6 max_buffered_rows = 4096;
```

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 4]
```diff
@@ -0,0 +1,762 @@
...
+
+    PFunctionCallRequest merge_buffer_request(PFunctionCallRequest& request) {
+        int args_size = _buffer_request[0].args_size();
+        request.set_function_name(_update_fn);
+        for (int i = 0; i < args_size; i++) {
+            PValues* arg = request.add_args();
+            arg->mutable_type()->CopyFrom(_buffer_request[0].args(i).type());
+            for (int j = 0; j < _buffer_request.size(); j++) {
+                for (int m = 0; m < _buffer_request[j].args(i).double_value_size(); m++) {
+                    arg->add_double_value(_buffer_request[j].args(i).double_value(0));
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:578]: what is m used for?

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 5]
```diff
@@ -0,0 +1,762 @@
...
+                }
+                for (int m = 0; m < _buffer_request[j].args(i).float_value_size(); m++) {
+                    arg->add_float_value(_buffer_request[j].args(i).float_value(0));
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:581]: There may be a more efficient way than adding one by one

**Comment by chenlinzhong** [comment]
> please add example to docs

done

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+constexpr int64_t max_buffered_rows = 4096;
+struct AggregateRpcUdafData {
+private:
+    std::string _update_fn;
+    std::string _merge_fn;
+    std::string _server_addr;
+    std::string _finalize_fn;
+    bool saved_last_result;
+    std::shared_ptr<PFunctionService_Stub> _client;
+    PFunctionCallResponse res;
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:55]:
```suggestion
    PFunctionCallResponse _res;
```

File: be/src/vec/aggregate_functions/aggregate_function_rpc.h [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+            } else {
+                convert_col_to_pvalue<false>(columns[i]->convert_to_full_column_if_const(),
+                                             argument_types[i], arg, start, end);
+            }
+        }
+        return Status::OK();
+    }
+
+#ifndef RPC_ADD
+#define RPC_ADD
```
Comment (yangzhg) [inline be/src/vec/aggregate_functions/aggregate_function_rpc.h:151]: no use？

File: docs/zh-CN/docs/ecosystem/udaf/remote-user-defined-aggregation-function.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+* 灵活：UDAF Service 中可以调用任意其他服务或程序库类，以满足更多样的业务需求。
+
+  
+
+2. 使用限制
+
+* 性能：相比于 Native UDAF，UDAF Service 会带来额外的网络开销，因此性能会远低于 Native UDAF。同时，UDAF Service 自身的实现也会影响函数的执行效率，用户需要自行处理高并发、线程安全等问题。
+
+* 单行模式和批处理模式：Doris 原先的基于行存的查询执行框架会对每一行数据执行一次 UDAF RPC 调用，因此执行效率非常差，而在新的向量化执行框架下，会对每一批数据（默认2048行）执行一次 UDAF RPC 调用，因此性能有明显提升。实际测试中，基于向量化和批处理方式的 Remote UDAF 性能和基于行存的 Native UDAF 性能相当，可供参考。
```
Comment (yangzhg) [inline docs/zh-CN/docs/ecosystem/udaf/remote-user-defined-aggregation-function.md:71]:
```suggestion
* 单行模式和批处理模式：Doris 原先的基于行存的查询执行框架会对每一行数据执行一次 UDAF RPC 调用，因此执行效率非常差，而在新的向量化执行框架下，会对每一批数据执行一次 UDAF RPC 调用，因此性能有明显提升。实际测试中，基于向量化和批处理方式的 Remote UDAF 性能和基于行存的 Native UDAF 性能相当，可供参考。
```

**Review by yangzhg — DISMISSED** [review]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**